### DISCUSSION TITLE
Allocation hotspot in S3ChecksumValidatingPublisher due to buffer copying

### DISCUSSION THREAD

**Discussion by ldaley** [aws/aws-sdk-java-v2#5086] [Q&A]
Hi,

While profiling an application reading a non-trivial amount of data from S3, I noticed that the S3 checksum validation is a significant allocation hotspot. The following method in particular:

https://github.com/aws/aws-sdk-java-v2/blob/master/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/checksums/S3ChecksumValidatingPublisher.java#L81

On first read, it seems that these buffer copies could be avoided by working directly with ByteBuffer. Before I start experimenting with a patch along those lines, I'd like to know whether it's a “policy” requirement to use discrete buffers at each step in the publisher/subscriber chain, or whether a patch that avoids copies here could be accepted.

TIA

**Comment by debora-ito** [comment]
Can you expand and share your findings on the "significant allocation hotspot", like the profiling reports?
